lib/protobuf/text.ex

```diff
@@ -0,0 +1,155 @@
+          ])
+        end
+
+      nil ->
+        if Enum.any?(message_props.oneof, fn {oneof_name, _} -> name == oneof_name end) do
+          case value do
+            {field_name, field_value} ->
+              encode_struct_field({field_name, field_value}, syntax, message_props)
+
+            nil ->
+              Algebra.empty()
+
+            _ ->
+              raise "invalid value for oneof `#{inspect(name)}`: #{inspect(value)}"
+          end
+        else
+          raise "unknown field #{inspect(name)}"
+        end
+    end
+  end
+
+  @spec encode_value(term(), :proto2 | :proto3, FieldProps.t()) :: Algebra.t()
+  defp encode_value(value, syntax, %{repeated?: true} = field_prop) when is_list(value) do
+    fun = fn val, _opts -> encode_value(val, syntax, field_prop) end
+    Algebra.container_doc("[", value, "]", inspect_opts(), fun, break: :strict)
+  end
+
+  defp encode_value(value, syntax, %{map?: true, repeated?: false} = field_prop) do
+    as_list =
+      Enum.map(value, fn {k, v} -> struct(field_prop.type, key: k, value: v) end)
+
+    encode_value(as_list, syntax, %{field_prop | repeated?: true})
+  end
+
+  defp encode_value(value, _syntax, %{embedded?: true, type: mod}) do
+    value
+    |> transform_module(mod)
+    |> encode_struct(mod.__message_props__())
+  end
+
+  defp encode_value(nil, :proto2, %FieldProps{required?: true, name_atom: name}) do
+    raise "field #{inspect(name)} is required"
+  end
+
+  defp encode_value(value, _, _) when is_atom(value) do
+    to_string(value)
+  end
+
+  defp encode_value(value, _, _) do
+    inspect(value)
+  end
+
+  defp transform_module(message, module) do
+    if transform_module = module.transform_module() do
+      transform_module.encode(message, module)
+    else
+      message
+    end
+  end
+
+  # Copied from Protobuf.Encoder. Should it be shared?
+  defp skip_field?(_syntax, [], _prop), do: true
+  defp skip_field?(_syntax, val, _prop) when is_map(val), do: map_size(val) == 0
+  defp skip_field?(:proto2, nil, %FieldProps{optional?: optional?}), do: optional?
+  defp skip_field?(:proto2, value, %FieldProps{default: value, oneof: nil}), do: true
+  defp skip_field?(:proto3, val, %FieldProps{proto3_optional?: true}), do: is_nil(val)
+
+  defp skip_field?(:proto3, nil, _prop), do: true
+  defp skip_field?(:proto3, 0, %FieldProps{oneof: nil}), do: true
+  defp skip_field?(:proto3, +0.0, %FieldProps{oneof: nil}), do: true
+  defp skip_field?(:proto3, "", %FieldProps{oneof: nil}), do: true
+  defp skip_field?(:proto3, false, %FieldProps{oneof: nil}), do: true
+
+  # This is actually new. Should it be ported to Protobuf.Encoder?
+  defp skip_field?(:proto3, value, %FieldProps{type: {:enum, enum_mod}, oneof: nil}) do
+    enum_props = enum_mod.__message_props__()
+    [first_tag | _] = enum_props.ordered_tags
+    %{name_atom: name_atom, fnum: fnum} = enum_props.field_props[first_tag]
+    value == name_atom or value == fnum
+  end
```

Comment:
I tried this on Protobuf.Encoder and it broke conformance tests. Despite the fact that in the docs it seems like this is the expected behaviour there as well (just as we skip `0` for non-optional integers, I'd expect it to skip the default enum value...)

Reply:
I'll investigate this further